src/System.Text.Json/src/System/Text/Json/JsonParseObject.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Text.Utf8;
+
+namespace System.Text.Json
+{
+    public struct JsonParseObject
+    {
+        private byte[] _buffer;
+        private int _start;
+        private int _end;
+
+        private const int RowSize = 9;
+
+        public JsonParseObject(byte[] buffer, int start, int end)
+        {
+            _buffer = buffer;
+            _start = start;
+            _end = end;
+        }
+
+        public bool HasValue()
+        {
+            var typeCode = _buffer[_start + 8];
+
+            if (typeCode == (byte)JsonParser.JsonValueType.Object || typeCode == (byte)JsonParser.JsonValueType.Array)
+            {
+                int length = BitConverter.ToInt32(_buffer, _start + 4);
+                if (length == 0) return false;
+                return true;
+            }
+            else
+            {
+                int location = BitConverter.ToInt32(_buffer, _start);
+                if (_buffer[location - 1] == '"' && _buffer[location + 4] == '"')
+                {
+                    return true;
+                }
+                return (_buffer[location] != 'n' || _buffer[location + 1] != 'u' || _buffer[location + 2] != 'l' || _buffer[location + 3] != 'l');
+            }
+        }
+
+        public JsonParseObject this[string index]
+        {
+            get
+            {
+                int length = BitConverter.ToInt32(_buffer, _start + 4);
+                var typeCode = _buffer[_start + 8];
+
+                if (length == 0)
+                {
+                    throw new KeyNotFoundException();
+                }
+
+                if (typeCode != (byte)JsonParser.JsonValueType.Object)
+                {
+                    throw new NullReferenceException();
```

Comment:
OK. So in the scenario when user searches for a key string within a Json Array, what should we do? Only integer indexing makes sense in that case.